packages/libs/error-stack/tests/snapshots/test_debug__linear@spantrace-glyph.snap

```diff
@@ -0,0 +1,19 @@
+---
+source: tests/test_debug.rs
+expression: "redact(&format!(\"{report:?}\"))"
+---
+Printable C
+│ tests/test_debug.rs:225:10
+├─▶ Context B
+│   ╰ tests/test_debug.rs:224:10
+├─▶ Printable B
+│   ╰ tests/test_debug.rs:222:10
+├─▶ Context A
+│   ╰ tests/test_debug.rs:221:10
+├─▶ Printable A
+│   ╰ tests/test_debug.rs:218:10
+├─▶ Root error
+│   ╰ tests/common.rs:140:5
+├─▶ spantrace with 2 frames (1)
+│   ╰ tests/common.rs:140:5
+╰─▶ 3 additional attachments
```

Comment:
mhhh okay, currently, we do not do any reordering of frames. Everything stays in the same order it was added in. I was thinking about reordering but unsure if that was wanted.

Would that mean that a change in context is also automatically an indentation? This is certainly doable, it just is a bit more complicated to implement.

Just for clarification @TimDiekmann, if I have the following linear stack: `Attachment A, Attachment B, Context C, Attachment D, Attachment E, Context F`, would I reorder the representation to basically render:

```
Context C 
  -> Attachment A
  -> Attachment B
  -> Context F 
    -> Attachment D
    -> Attachment E
```

so that I basically "bring" the context up?

What happens if I have a "join" at an attachment? So `Attachment A, Attachment B [(Attachment D, Context E), (Attachment F, Context E)]` How would that render? I wouldn't be able to "surface" a context for `Attachment A` and `Attachment B`, should I just use `Attachment A` as the top?

Reply:
We could also opt for an `Opaque Context` if that's the case and to stay consistent.

Reply:
> I was thinking about reordering but unsure if that was wanted.

I thought about this as well but this would introduce more issues than it solves. I guess by reordering you mean when ...
- ... attaching you would
    1. pop the context
    2. push the attachment
    3. push the context
- ... changing context you would
    1. push the context

When attaching if there are multiple top-level errors, to which chain the attachment would be added? My intuitive answer would be both, which would requires the attachment to be clonable.

---

> if I have the following linear stack: `Attachment A, Attachment B, Context C, Attachment D, Attachment E, Context F` ...

To avoid confusion, `Context F` is your `Context` you created the `Report` with? Then this would result in 

```
Context C 
| - Attachment B
| - Attachment A
-> Context F 
  - Attachment E
  - Attachment D
```

The [announcement blog post](https://hash.dev/blog/announcing-error-stack) has a similar example in the "Attachments" section:
```rust
fn parse_config(path: impl AsRef<Path>) -> Result<Config, Report<ParseConfigError>> {
    let path = path.as_ref();

    let content = fs::read_to_string(path)
        .report()
        .change_context(ParseConfigError::new())
        .attach_printable_lazy(|| format!("Could not read file {path:?}"))
        .attach(Suggestion("Use a file you can read next time!"))?;

    // ...
}
```

```
Could not parse configuration file
             at main.rs:6:10
      - Could not read file "config.json"
      - 1 additional opaque attachment

Caused by:
   0: No such file or directory (os error 2)
             at main.rs:5:10
```

This results in a natural way of writing the code. If you change the context, then add attachments, the attachment you added first will also appear first.

---

> What happens if I have a "join" at an attachment?

Could you show me an example of this as code? Then I think it's easier to construct an output from that 🙂

Reply:
> I thought about this as well but this would introduce more issues than it solves. I guess by reordering you mean when ...

I think you misunderstood me. This is only during rendering in `Debug`, not during changing the frame stack, through `attach()` etc. What I mean is:

```text
Attachment A
Attachment B
Context C
Attachment D
Attachment E
Context F
```

During rendering reorder to:

```text
Context C
  Attachment A
  Attachment B
Context F
  Attachment D
  Attachment E
```

basically "lifting" the context.

--- 

> To avoid confusion, `Context F` is your `Context` you created the `Report` with? Then this would result in

Yes, that's correct. This is the output that I was expecting too. Great that we're on the same page. I have some ideas about adjusting the output to reflect this correctly (I will post once ready).

----

> Could you show me an example of this as code? Then I think it's easier to construct an output from that 🙂

```rust
    let mut report = create_report().attach_printable(PrintableA(0));
    report.extend_one({
        let mut report = create_report().attach_printable(PrintableB(1));
    });

    let report = report.attach(AttachmentA(2)).attach(AttachmentB(2));
```

This is what I meant. This creates the following tree:

```mermaid
flowchart BT;

RootError1 --> PrintableA
RootError2 --> PrintableB

PrintableA --> AttachmentA
PrintableB --> AttachmentA

AttachmentA --> AttachmentB
```

where `AttachmentA`, `AttachmentB` has no "previous" `Context`. In this case, either `AttachmentA` would be the "top," or we print something like "Opaque Context"

Reply:
This example a bit more complex: 

```mermaid
flowchart BT;

RootError1 --> PrintableA
PrintableA --> ContextC
ContextC --> PrintableD
RootError2 --> PrintableB

PrintableD --> AttachmentA
PrintableB --> AttachmentA

AttachmentA --> AttachmentB

ContextE --> AttachmentF
AttachmentF --> AttachmentG

AttachmentB --> ContextE
```

would approximately render into something like this:
(I am still unsure about the error heads.

```
Context E
├╴(location)
├╴Attachment G
├╴Attachment F
├─◈ Opaque Context
│   ├╴(location of Attachment B)
│   ├╴Attachment B
│   ├╴Attachment A
├─▶ Context A
│   ├╴(location)
│   ├╴Printable A
│   ╰─◈ RootError 1
│       ├╴(location)
│       ╰╴Printable A
╰─▶ RootError 2
    ├╴(location)
    ╰╴Printable B
```

Reply:
I would probably duplicate the output for the context, so for

> ```mermaid
> flowchart BT;
>
> RootError1 --> PrintableA
> RootError2 --> PrintableB
>
> PrintableA --> AttachmentA
> PrintableB --> AttachmentA
>
> AttachmentA --> AttachmentB
> ```

this would become

```
RootError1
   ├ Printable A
   ├ AttachmentA
   ╰ AttachmentB

RootError2
   ├ Printable B
   ├ AttachmentA
   ╰ AttachmentB
```

and for 

```mermaid
flowchart BT;

RootError1 --> PrintableA
PrintableA --> ContextC
ContextC --> PrintableD
RootError2 --> PrintableB

PrintableD --> AttachmentA
PrintableB --> AttachmentA

AttachmentA --> AttachmentB

ContextE --> AttachmentF
AttachmentF --> AttachmentG

AttachmentB --> ContextE
```

this would be

```
Context E
│   ├╴Attachment F
│   ╰╴Attachment G
├─▶ Context C
│   │   ├╴Printable D
│   │   ├╴AttachmentA
│   │   ╰╴AttachmentB
│   ╰─▶ RootError 2
│       ╰╴Printable B
╰─▶ RootError 1
    ├╴Printable A
    ├╴AttachmentA
    ╰╴AttachmentB
```

I like the Arrow and also tried to make it readable, what do you think of the output? cc @hashintel/error-stack 

Maybe also a newline before contexts improves 

```
Context E
│   ├╴Attachment F
│   ╰╴Attachment G
│
├─▶ Context C
│   │   ├╴Printable D
│   │   ├╴AttachmentA
│   │   ╰╴AttachmentB
│   │
│   ╰─▶ RootError 2
│       ╰╴Printable B
│
╰─▶ RootError 1
    ├╴Printable A
    ├╴AttachmentA
    ╰╴AttachmentB
```

Reply:
> I would probably duplicate the output for the context, so for

I like this one!

I think the arrow (▶) is fine. I was experimenting with ◆ (and others: https://www.w3.org/TR/xml-entity-names/025.html) to indicate a difference between split vs. new context. I thought using ◆ (<> in ASCII)  for a context and ▶ (-> in ASCII) for a new group might be fitting.

> Maybe also a newline before contexts improves readability

100%

---

I don't know if I am fan of the extra indentation in:

```
├─▶ Context C
│   │   ├╴Printable D
│   │   ╰╴AttachmentA
│   │
│   ╰─▶ RootError 2
``` 

@TimDiekmann do you think


```
├─▶ Context C
│   ├╴Printable D
│   ├╴AttachmentA
│   │
│   ╰─▶ Context D
```

wouldn't be readable enough? I am trying to minimize indentation. Otherwise, we might end up with something exploding horizontally pretty quickly.

I think that overall they accomplish the same thing, except that we would need an additional special case in you case that might lead to weirdness.

```
│   ╰─▶ RootError 2
│       ╰╴Printable B
```

vs.

```
├─▶ Context C
│   │   ├╴Printable D
```

which might confuse users too, if some have an extra indent

Reply:
> ```
> ├─▶ Context C
> │   ├╴Printable D
> │   ├╴AttachmentA
> │   │
> │   ╰─▶ Context D
> ```

You are absolutely right, with the extra new line this is far more readable and sufficient! 🙂

Reply:
I think this would then be the full example?
```
Context E
├╴Attachment F
├╴Attachment G
│
├─▶ Context C
│   ├╴Printable D
│   ├╴AttachmentA
│   ├╴AttachmentB
│   │
│   ╰─▶ RootError 2
│       ╰╴Printable B
│
╰─▶ RootError 1
    ├╴Printable A
    ├╴AttachmentA
    ╰╴AttachmentB
```

Reply:
> I think the arrow (`▶`) is fine. I was experimenting with ◆ (and others: https://www.w3.org/TR/xml-entity-names/025.html) to indicate a difference between split vs. new context. I thought using `◆` (`<>` in ASCII) for a context and `▶` (`->` in ASCII) for a new group might be fitting.

I think I like the arrow the most, but don't have a strong opinion on that. Especially, as I like the `->` output more than `<>` 😄

Reply:
Yes, LGTM. Then I will change the current logic for rendering to be able to represent this if that's ok w/ everyone.

> I think I like the arrow the most, but don't have a strong opinion on that. Especially, as I like the -> output more than <> 😄

Same, I am just a bit afraid that overwise, we lose information about: what is a split, what is a context? This might not be a big deal IRL, but that is why I am considering different arrows for both.

(This isn't a blocker, due to the instruction based architecture this is likely just a single line change down the line)

Reply:
Then I'd say we can move on with this. I will do another deep code review when this was added, so we can unblock the 0.2 release. The most important thing IMO is that we have an output corresponding to the actual structure, not a flat one as we do currently. 🚀

Reply:
One thought on `"{:?}"` vs `"{:#?}"`: I would probably output the location and information on both, but instead of the `Display` implementation we could use the `Debug` implementation for the alternate form. This would at least give more value.

Reply:
What do you think rendering for more than two switches would look like? e.g.

```mermaid
flowchart BT;

RootError1 --> PrintableA
RootError2 --> PrintableB
RootError3 --> PrintableC

PrintableA --> AttachmentA
PrintableB --> AttachmentA
PrintableC --> AttachmentA

AttachmentA --> ContextA
ContextA --> AttachmentB
```

Maybe

```
Context A
├╴Attachment B
│
├─▶ RootError 3
│   │ ├╴Printable C
│   │ ╰╴AttachmentA
│   │
│   ╰─▶ RootError 2
│       ├╴Printable B
│       ╰╴AttachmentA
│
╰─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

Maybe, `◆` for splitting isn't that bad then:

```
Context A
├╴Attachment B
│
◆─◆─▶ RootError 3
│ │   ├╴Printable C
│ │   ╰╴AttachmentA
│ │
│ ╰─▶ RootError 2
│     ├╴Printable B
│     ╰╴AttachmentA
│
╰─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

Reply:
What about?

```
Context A
├╴Attachment B
│
╞═▶ RootError 3
│   ├╴Printable C
│   ╰╴AttachmentA
│
╞═▶ RootError 2
│   ├╴Printable B
│   ╰╴AttachmentA
│
╘═▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

or:

```
Context A
├╴Attachment B
│
◆
│
├─▶ RootError 3
│   ├╴Printable C
│   ╰╴AttachmentA
│
├─▶ RootError 2
│   ├╴Printable B
│   ╰╴AttachmentA
│
╰─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

or:

```
Context A
├╴Attachment B
│
▼
▲
│
├─▶ RootError 3
│   ├╴Printable C
│   ╰╴AttachmentA
│
├─▶ RootError 2
│   ├╴Printable B
│   ╰╴AttachmentA
│
╰─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

or?

```
Context A
├╴Attachment B
│
◊
│
├─▶ RootError 3
│   ├╴Printable C
│   ╰╴AttachmentA
│
├─▶ RootError 2
│   ├╴Printable B
│   ╰╴AttachmentA
│
╰─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```

or:

```
Context A
├╴Attachment B
│
│
◆─▶ RootError 3
│   ├╴Printable C
│   ╰╴AttachmentA
│
◆─▶ RootError 2
│   ├╴Printable B
│   ╰╴AttachmentA
│
◆─▶ RootError 1
    ├╴Printable A
    ╰╴AttachmentA
```


My problem kind of with all of them, is that there's a missing distinction between splits and contexts, I feel like none of them are ideal. These are just some of the things I have been thinking about.

> Maybe, `◆` for splitting isn't that bad then: [...]

my problem with that approach is that we end up w/ multiple intendations which differ slighlty, meaning we don't really get a nice cascade if it goes down any more. It also kind of looks `RootError 3`
 and `RootError 2` are related to each other and then `RootError 1`, which can make this confusing, but I very much like the direction it is going!